# pPAnnot: Parse genome annotations for functional genomic study

## 1. Introduction
The bioinformatics tool pPAnnot could create on-site genome annotations and parse them with datasets. The datasets could be read counts table determined by high-throughput sequence data. The parsed datasets could be used for functional genomic study, such as ontology enrichment, pathway enrichment, clustering, etc.

pPannot is very flexible because the parsed genome annotation is term-based. A specific genome annotations namely human or mouse could be parsed according to your study. Or, a certain gene-like annotations could be parsed. For example, those term-based annotations could be all prostate cancer related (oncology study), cellulase (microbio-environments study), and so on.

## 2. Testing and installation

### 2.1. unit test

### 2.2. local test

### 2.3. installation


## 3. Quick start

### step 1: download source data from NCBI and ExPASy


### step 2: build local database


### step 3: parse genome annotations


## 4. references
